File: 프로그래머스/unrated/132266. 부대복귀/부대복귀.py
import heapq

INF = int(1e9)


def solution(n, roads, sources, destination):
    answer = []
    graph = [[] for _ in range(n + 1)]
    distance = [INF] * (n + 1)
    
    for a, b in roads:
        graph[a].append((b, 1))
        graph[b].append((a, 1))
        
    q = []
    heapq.heappush(q, (0, destination))
    distance[destination] = 0
    
    while q:
        dist, now = heapq.heappop(q)
        if distance[now] < dist:
            continue
        for v, c in graph[now]:
            cost = dist + c
            if cost < distance[v]:
                distance[v] = cost
                heapq.heappush(q, (cost, v))
                
    for source in sources:
        if distance[source] == INF:
            answer.append(-1)
        else:
            answer.append(distance[source])
            
    return answer

File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {

    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        List<int[]>[] graph = new List[n + 1];
        for (int i = 0; i <= n; i++) graph[i] = new ArrayList<>();

        for (int[] road : roads) {
            graph[road[0]].add(new int[]{road[1], 1});
            graph[road[1]].add(new int[]{road[0], 1});
        }

        int[] dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[destination] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, destination});

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int d = cur[0], now = cur[1];

            if (dist[now] < d) continue;

            for (int[] next : graph[now]) {
                int cost = d + next[1];
                if (cost < dist[next[0]]) {
                    dist[next[0]] = cost;
                    pq.offer(new int[]{cost, next[0]});
                }
            }
        }

        int[] answer = new int[sources.length];
        for (int i = 0; i < sources.length; i++) {
            answer[i] = dist[sources[i]] == Integer.MAX_VALUE ? -1 : dist[sources[i]];
        }
        return answer;
    }
}